### ISSUE TITLE
middleware bundle downloader fails w/ "otto" driver

### ISSUE THREAD

**Issue by bitsofinfo** [TykTechnologies/tyk#992]
**Do you want to request a *feature* or report a *bug*?**

bug

**What is the current behavior?**

In my gateway's (2.3.7)  tyk.conf the following is setup

```
    "enable_bundle_downloader": true,
    "bundle_base_url": "http://192.168.0.216:8888/tyk/",
    "coprocess_options": {
       "enable_coprocess": true,
       "coprocess_grpc_server": ""
    },
```
My API def has the following: 

    "custom_middleware_bundle": "my-bundle.zip",

I can download manually in a browser: http://192.168.0.216:8888/tyk/my-bundle.zip (its unsigned)

I next restart the gateway and get this in the logs:

```
Couldn't create bundle directory: mkdir middleware/bundles/ba2c643525a54dc34b975179a7d9f12a-my-bundle.zip: no such file or directory" api_id=ba2c643525a54dc34b975179a7d9f12a org_id=5976507eb46f2e00016bb4a1 path=- server_name="http://192.168.0.216:9080/mytestapi" user_id=- user_ip=-
```

Next I login to the container and I manually created the `middleware/bundles` directory, restart the gateway and now I get this

```
2017-08-07T14:58:32.776748541Z time="Aug 7 14:58:32" level=info msg="----> Loading bundle: my-bundle.zip"
2017-08-07T14:58:32.776788593Z time="Aug 7 14:58:32" level=info msg="----> Verifying bundle: my-bundle.zip"
2017-08-07T14:58:32.777200804Z time="Aug 7 14:58:32" level=info msg="----> Bundle is valid, adding to spec: my-bundle.zip"
2017-08-07T14:58:32.777237061Z time="Aug 7 14:58:32" level=error msg="Failed to load Middleware JS: read middleware/bundles/ba2c643525a54dc34b975179a7d9f12a-my-bundle.zip: is a directory"
2017-08-07T14:58:32.777260533Z time="Aug 7 14:58:32" level=error msg="Failed to load Middleware JS: read middleware/bundles/ba2c643525a54dc34b975179a7d9f12a-my-bundle.zip: is a directory
```

If I shell into the gateway I do see this directory exists and is legit w/ the contents of the zip file (my js manifest and middleware)

```
/opt/tyk-gateway/middleware/bundles/ba2c643525a54dc34b975179a7d9f12a-my-bundle.zip
```

**What is the expected behavior?**

That it should not have there errors, nor should I have to manually create the `bundles` directory

**Which versions of Tyk affected by this issue? Did this work in previous versions of Tyk?**

2.3.7/1.3.7

**Comment by matiasinsaurralde** [comment]
I've been able to reproduce this, the main issue is that the `LoadJSPaths`expects the path to be set.
When a JS manifest is prepared and a bundle is generated, the manifest looks like this:
```json
{
  "file_list": [
    "myMiddleware1.js",
    "myMiddleware2.js"
  ],
  "custom_middleware": {
    "pre": [
      {
        "name": "myMiddleware1"
      },
      {
        "name": "myMiddleware2"
      }
    ],
    "driver": "otto",
  },
  "checksum": "",
  "signature": ""
}
```
When `LoadJSPaths` iterates it doesn't get the paths to the actual files and it breaks. I will send a patch soon.
As a workaround for users that are getting this error, they should append the path in their manifest files:
```json
{
  "file_list": [
    "myMiddleware1.js",
    "myMiddleware2.js"
  ],
  "custom_middleware": {
    "pre": [
      {
        "name": "myMiddleware1",
        "path": "myMiddleware1.js"
      },
      {
        "name": "myMiddleware2",
        "path": "myMiddleware2.js"
      }
    ],
    "driver": "otto",
  },
  "checksum": "",
  "signature": ""
}
```
This looks like a repetitive thing, however it's important to note that the plugin loader doesn't access `file_list`, and `file_list` is just used by the CLI tool, this is why the path is needed inside the `custom_middleware` block. I don't think the user should manually write the manifest in this way but we should ensure the path gets injected on runtime.

**Comment by bitsofinfo** [comment]
Confirmed the workaround of listing the paths fixes it.

**Comment by bitsofinfo** [comment]
Note this error still exists: `Couldn't create bundle directory: mkdir middleware/bundles/c9bd7307806b495c7a5971c72087f75f-my-bundle.zip`

I still need to manually go into the gateway and create that bundles directory...

**Comment by matiasinsaurralde** [comment]
Hi, do you already have a `middleware/bundles` directory?

**Comment by bitsofinfo** [comment]
No, we have to manually create it in order for the downloading to work

**Comment by bitsofinfo** [comment]
Is this fixed now? should this be closed via some commit int 2.3.10?

**Comment by ilijabojanovic** [comment]
@bitsofinfo This is fixed in 2.3.10 release
